pycloudlib/ec2/instance.py

```diff
@@ -379,12 +466,61 @@ def _get_nic_matching_ip(self, ip_address):
             None,
         )
 
+    def _find_nic_associations(self, nics_resp: Dict) -> List[Dict]:
+        associations: List[Dict] = []
+        if not nics_resp["NetworkInterfaces"]:
+            return associations
+        for nic_md in nics_resp["NetworkInterfaces"]:
+            # Global associations
+            association = nic_md.get("Association")
+            if association:
+                associations.append(association)
+            # Per-ip associations
+            for priv_address in nic_md.get("PrivateIpAddresses", []):
+                if priv_address.get("Primary", False):
+                    continue
+                association = priv_address.get("Association")
+                if association:
+                    associations.append(association)
+        return associations
+
+    def _release_address(self, private_ip: str):
+        nics_resp = self._client.describe_network_interfaces(
+            Filters=[
+                {
+                    "Name": "private-ip-address",
+                    "Values": [private_ip],
+                },
+            ]
+        )
+        for association in self._find_nic_associations(nics_resp):
+            try:
+                self._client.disassociate_address(
```

Comment:
Do we even need to separately call disassociate_address? [Botocore seems to indicate that release_address will automatically disassociate](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/ec2/client/release_address.html#release-address)

Reply:
That's only true in the default VPC:

> [Default VPC] Releasing an Elastic IP address automatically disassociates it from any instance that it’s associated with. To disassociate an Elastic IP address without releasing it, use DisassociateAddress.

> [Nondefault VPC] You must use DisassociateAddress to disassociate the Elastic IP address before you can release it. Otherwise, Amazon EC2 returns an error ( InvalidIPAddress.InUse).

Reply:
Thanks for this @aciba90